apps/routing_transparency.go

```diff
@@ -33,17 +34,47 @@ var _ = AppsDescribe("Routing Transparency", func() {
 		Expect(cf.Cf("delete", appName, "-f", "-r").Wait()).To(Exit(0))
 	})
 
-	It("Supports URLs with percent-encoded characters", func() {
-		curlResponse := helpers.CurlApp(Config, appName, "/requesturi/%21%7E%5E%24%20%27%28%29?foo=bar+baz%20bing")
-		Expect(curlResponse).To(ContainSubstring("Request"))
-		Expect(curlResponse).To(ContainSubstring("/requesturi/%21%7E%5E%24%20%27%28%29"))
-		Expect(curlResponse).To(ContainSubstring("Query String is [foo=bar+baz%20bing]"))
+	Context("Kubernetes Istio/Envoy Proxy Behavior", func() {
+		BeforeEach(func() {
+			if !Config.GetUseKubernetesIstioProxy() {
+				Skip(skip_messages.SkipKubernetesNetworking)
+			}
+		})
+
+		It("normalizes URLs with percent-encoded characters by decoding certain characters", func() {
+			curlResponse := helpers.CurlApp(Config, appName, "/requesturi/%21%7E%5E%24%20%27%28%29?foo=bar+baz%20bing")
+			Expect(curlResponse).To(ContainSubstring("Request"))
+			Expect(curlResponse).To(ContainSubstring("/requesturi/%21~%5E%24%20%27%28%29"))
+			Expect(curlResponse).To(ContainSubstring("Query String is [foo=bar+baz%20bing]"))
+		})
+
+		It("encodes certain reserved/unsafe characters", func() {
+			curlResponse := helpers.CurlApp(Config, appName, "/requesturi/!~^'()$\"?!'()$#!'")
+			Expect(curlResponse).To(ContainSubstring("Request"))
+			Expect(curlResponse).To(ContainSubstring("/requesturi/!~%5E'()$%22"))
+			Expect(curlResponse).To(ContainSubstring("Query String is [!'()$]"))
+		})
 	})
 
-	It("transparently proxies both reserved characters and unsafe characters", func() {
-		curlResponse := helpers.CurlApp(Config, appName, "/requesturi/!~^'()$\"?!'()$#!'")
-		Expect(curlResponse).To(ContainSubstring("Request"))
-		Expect(curlResponse).To(ContainSubstring("/requesturi/!~^'()$\""))
-		Expect(curlResponse).To(ContainSubstring("Query String is [!'()$]"))
+	Context("Gorouter Behavior", func() {
+		BeforeEach(func() {
+			if Config.GetUseKubernetesIstioProxy() {
+				Skip(skip_messages.SkipGorouterSpecificBehavior)
+			}
+		})
+
+		It("supports URLs with percent-encoded characters", func() {
+			curlResponse := helpers.CurlApp(Config, appName, "/requesturi/%21%7E%5E%24%20%27%28%29?foo=bar+baz%20bing")
+			Expect(curlResponse).To(ContainSubstring("Request"))
+			Expect(curlResponse).To(ContainSubstring("/requesturi/%21%7E%5E%24%20%27%28%29"))
+			Expect(curlResponse).To(ContainSubstring("Query String is [foo=bar+baz%20bing]"))
+		})
+
+		It("transparently proxies both reserved characters and unsafe characters", func() {
+			curlResponse := helpers.CurlApp(Config, appName, "/requesturi/!~^'()$\"?!'()$#!'")
+			Expect(curlResponse).To(ContainSubstring("Request"))
+			Expect(curlResponse).To(ContainSubstring("/requesturi/!~^'()$\""))
+			Expect(curlResponse).To(ContainSubstring("Query String is [!'()$]"))
+		})
```

Comment:
It took me a while to see how the two separate blocks differ.  I would have preferred something like this:
```
var tildeRepresentation string
if Config.GetUseKubernetesIstioProxy() {
  tildeRepresentation = "~"
} else {
  tildeRepresentation = "%7E"
}
It("normalizes URLs with percent-encoded characters by decoding certain characters", func() {
	curlResponse := helpers.CurlApp(Config, appName, "/requesturi/%21%7E%5E%24%20%27%28%29?foo=bar+baz%20bing")
	Expect(curlResponse).To(ContainSubstring("Request"))
	Expect(curlResponse).To(ContainSubstring("/requesturi/%21"  + tildeRepresentation + "%5E%24%20%27%28%29"))
	Expect(curlResponse).To(ContainSubstring("Query String is [foo=bar+baz%20bing]"))
})
...
```

Reply:
I understand that it's hard to see the difference, but these tests are not meant for comparison of behaviors rather than testing proxies' routing transparencies. 

The difference can be found in the test message titles. Gorouter tests assuming that everything is passed without any kind of normalization, while Istio tests assume that some unsafe characters are normalized. We'll few `By` messages to describe the assertions the tests are checking for to make this clear.